Fix various things -fsanitize complains about

sipa opened the pull request:
* The ReadLE32 etc functions in crypto/common.h where dereferencing type-punned pointers, which is undefined behaviour (and can cause consistent failure on some platforms due to misaligned reads). Fix this by `memcpy()`ing into a local variable first (as accessing a type's representation through a char* is always allowed).
* The scriptnum tests were testing running several arithmetic operations on large integers, triggering signed integer overflow, which is undefined. Fix the tests by using integers limited to +/- 2^40 (larger than anything actually supported in script).
* Fix a memory leak in the wallet tests: CWalletTx objects that weren't freed - wrap them in a unique_ptr.
* Fix a memory leak in the net tests: a CNode object wasn't being freed - wrap them in a unique_ptr.
* Fix a memory leak in the rpc auth: a buffer that wasn't being freed - switch it to a stack-allocated array.
* The REST code for getutxos was using boost::dynamic_bitset, which apparently issues a rightshift by a negative amount. Fix this by replacing it with normal bit/byte vectors.

Now all unit tests and rpc tests run succesfully with -fsanitize=address -fsanitize=undefined -fsanitize=leak in GCC 6.2.0.

Thanks to @kcc for pointing me to these tools.

sipa committed:
Avoid unaligned access in crypto i/o

sipa committed:
Avoid integer overflows in scriptnum tests

sipa committed:
Fix memory leak in wallet tests

sipa committed:
Fix memory leak in net_tests

sipa committed:
Fix memory leak in multiUserAuthorized

dcousens commented on src/test/scriptnum_tests.cpp:
```diff
@@ -13,7 +13,7 @@
 
 static const int64_t values[] = \
-{ 0, 1, CHAR_MIN, CHAR_MAX, UCHAR_MAX, SHRT_MIN, USHRT_MAX, INT_MIN, INT_MAX, UINT_MAX, LONG_MIN, LONG_MAX };
+{ 0, 1, CHAR_MIN, CHAR_MAX, UCHAR_MAX, SHRT_MIN, USHRT_MAX, INT_MIN, INT_MAX, UINT_MAX, (1LL << 40), (1LL << 40) - 1, -(1LL << 40), 1 - (1LL << 40)};
```
maybe define these constants ourselves rather than inlining them?

laanwj replied:
What is the problem that sanitize reports here? I don't understand the rationale for using "40" here.

sipa replied:
I've just replaced them with integer constants and added a comment.

sipa replied:
Overflow of signed integers is undefined behaviour. These numbers are added/subtracted from eachother, so that behaviour triggers. However, it doesn't matter in practice, as we only use scriptnums up to 4 bytes (which can turn into 5 bytes after arithmetic).

laanwj commented on src/crypto/common.h:
```diff
@@ -5,62 +5,98 @@
 uint64_t static inline ReadLE64(const unsigned char* ptr)
 {
-    return le64toh(*((uint64_t*)ptr));
+    return ((uint64_t)(ptr[0])) |
```
Interesting. This would also have failed on CPUs that don't support non-aligned reads (though I know of no such instances).

sipa replied:
Seems ARM supports unaligned reads pretty much since the architecture we support.

sipa commented:
I measure a 1.5% slowdown from these changes for the SHA256_32b benchmark (i7-6820HQ CPU @ 2.70GHz).

JeremyRubin commented on src/crypto/common.h:
```diff
@@ -5,62 +5,98 @@
 {
-    *((uint16_t*)ptr) = htole16(x);
+    ptr[0] = x & 0xFF;
+    ptr[1] = (x >> 8) & 0xFF;
```
I think the & 0xFF is redundant here, no? Unless you think there is some pipelining advantage...

JeremyRubin commented on src/crypto/common.h:
```diff
@@ -5,62 +5,98 @@
+    ptr[0] = x & 0xFF;
+    ptr[1] = (x >> 8) & 0xFF;
+    ptr[2] = (x >> 16) & 0xFF;
+    ptr[3] = (x >> 24) & 0xFF;
```
I think the & 0xFF is redundant here, no? Unless you think there is some pipelining advantage...

JeremyRubin commented on src/crypto/common.h:
```diff
@@ -5,62 +5,98 @@
+    ptr[4] = (x >> 32) & 0xFF;
+    ptr[5] = (x >> 40) & 0xFF;
+    ptr[6] = (x >> 48) & 0xFF;
+    ptr[7] = (x >> 56) & 0xFF;
```
I think the & 0xFF is redundant here, no? Unless you think there is some pipelining advantage...

JeremyRubin commented on src/crypto/common.h:
```diff
@@ -5,62 +5,98 @@
 void static inline WriteBE32(unsigned char* ptr, uint32_t x)
 {
-    *((uint32_t*)ptr) = htobe32(x);
+    ptr[0] = (x >> 24) & 0xFF;
```
I think the & 0xFF is redundant here, no? Unless you think there is some pipelining advantage...

JeremyRubin commented on src/crypto/common.h:
```diff
@@ -5,62 +5,98 @@
 void static inline WriteBE64(unsigned char* ptr, uint64_t x)
 {
-    *((uint64_t*)ptr) = htobe64(x);
+    ptr[0] = (x >> 56) & 0xFF;
```
I think the & 0xFF is redundant here, no? Unless you think there is some pipelining advantage...

JeremyRubin reviewed (approved):
utACK.

It seems there are some extra bitmasks in the write procedures; I've marked where they are. Maybe the optimized missed these and that's your 1.5% ;)

gmaxwell commented:
@sipa it's frustrating that the compiler doesn't recognize that pattern and peephole optimize it on platforms that support unaligned reads.  At the same time, we really should be using platform optimized SIMD sha256 on our primary platforms, so the performance of this code won't matter once we do.

laanwj commented:
Specifying &0xff for an unsigned char operation will just be ignored by the compiler (this is a no-brainer optimization). I doubt that is the cause of the slowdown.

I wonder, isn't there some compiler intrinsic for "non-aligned read of larger type" that could be detected by autoconf and used if available?

sipa committed:
Avoid boost dynamic_bitset in rest_getutxos

sipa commented:
Updated to use a different approach to avoiding unaligned access, which seems as fast as master for the SHA256_32b benchmark (or even faster, I consistently across different runs see a minimum duration speedup of 0.02-0.025%).

I've also added a commit to remove the use of boost::dynamic_bitset (which ubsan complains about, though I can't see exactly how or why).

Now all unit tests and rpc tests run succesfully with `-fsanitize=address -fsanitize=undefined -fsanitize=leak` in GCC 6.2.0.

jtimon commented:
utACK individual commits: 5a0b7e4106bc97a7a67bda6bf6fbd7f26d892420

Note: I will edit this post as I go, the list with 1 will hopefully grow, maybe up to 6

gmaxwell commented:
utACK.

gmaxwell commented:
ACK.

laanwj commented:
utACK 82e8baa
